tests/integration/common/s2n_test_common.py

```diff
@@ -0,0 +1,135 @@
+#
+#  http://aws.amazon.com/apache2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+
+"""
+Common functions to run s2n integration tests.
+"""
+
+import subprocess
+
+from common.s2n_test_scenario import Mode, Version, run_scenarios
+from common.s2n_test_reporting import Result
+from s2n_test_constants import TEST_ECDSA_CERT, TEST_ECDSA_KEY
+
+
+def get_error(process):
+    return process.stderr.readline().decode("utf-8")
+
+
+def wait_for_output(process, marker, line_limit=10):
+    for count in range(line_limit):
+        line = process.stdout.readline().decode("utf-8")
+        if marker in line:
+            return True
+    return False
+
+
+def cleanup_processes(*processes):
+    for p in filter(None, processes):
+        p.kill()
+        p.wait()
+
+
+def get_process(cmd):
+    return subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+
+def connect(get_peer, scenario):
+    """
+    Perform a handshake between s2n and another TLS process.
+
+    Args:
+        get_peer: a function that returns a TLS process
+        scenario: the handshake configuration
+
+    Returns:
+        A tuple of (client, server) processes which have completed a handshake.
+
+    """
+    server = get_s2n(scenario) if scenario.s2n_mode.is_server() else get_peer(scenario)
+    client = get_peer(scenario) if scenario.s2n_mode.is_server() else get_s2n(scenario)
+
+    return (server, client)
+
+
+def run_connection_test(get_peer, scenarios, test_func=None):
+    """
+    For each scenarios, s2n will attempt to perform a handshake with another TLS process
+    and then run the given test.
+
+    Args:
+        get_peer: a function that returns a TLS process for s2n to communicate with.
+            For an example of how to use this argument, see s2n_test_openssl.py.
+        scenarios: a list of handshake configurations
+        test_func: a function that takes a server and client process, performs some
+            additional testing, and then returns a Result
+
+    Returns:
+        A Result object indicating success or failure. If the given test returns
+        false or either the connection or the test throw an AssertionError, the Result
+        will indicate failure.
+
+    """
+    def __test(scenario):
+        client = None
+        server = None
+        try:
+            server, client = connect(get_peer, scenario)
+            if test_func:
+                return test_func(server, client)
+            else:
+                return Result()
+        except AssertionError as error:
+            return Result(error)
+        finally:
+            cleanup_processes(server, client)
+
+    return run_scenarios(__test, scenarios)
+
+
+def get_s2n_cmd(scenario):
```

Comment:
since Scenario is a class, would it make more sense for this function to be defined as part of the class's function? similarly for get_openssl_cmd()

Reply:
I initially started with that structure (OpensslScenario subclassed Scenario), but it didn't seem to provide any benefit and I'm not sure it really made sense from a separation of responsibility standpoint.